From 5207517143c9a9e2894435377adc32ad544d05f5 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Tue, 30 Sep 2025 08:15:49 +0200
Subject: [PATCH] media: stm32: dcmipp: fix default histogram configuration

Update the default histogram configuration in order to ensure
it is always valid in order to avoid getting bad configuration
error when starting a stream.

Fixes: a37a7a8bee10 ("media: stm32: dcmipp: add default histogram area settings in dcmipp-isp.c")'

Change-Id: I5dc0c3e4de2a594db7ed433b6ba2d45fa28a115b
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/497937
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        | 22 ++++++++++++++-----
 1 file changed, 17 insertions(+), 5 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -518,8 +518,19 @@ static void dcmipp_isp_config_decimation
 
 /* Histogram block - only available starting from stm32mp21 */
 #define DCMIPP_P1HSCR			0x8b0
-/* 4 Comp / 64 bins per comp / 1 region / decimated by 2*/
-#define DCMIPP_P1HSCR_DEFAULT		0x08411000
+/*
+ * Adopt a default Histogram configuration allowing to always work
+ * independently from the input format and resolution
+ * That is:
+ *   - after demosaicing
+ *   - 4 components
+ *   - 1 region only (full frame, rounded to decimation factors)
+ *   - 64 bins per comp (closest and lower to 320 / 4)
+ *   - Decimation, need a total of 5Mpix / 65536, that is around 76
+ *     - H decimation by 16
+ *     - V decimation by 8
+ */
+#define DCMIPP_P1HSCR_DEFAULT		0x08434006
 
 #define DCMIPP_P1HSSTR			0x8b4
 #define DCMIPP_P1HSSTR_START(x, y)	((x) | ((y) << 16))
@@ -539,10 +550,11 @@ static void dcmipp_isp_config_histo(stru
 	 * valid settings
 	 */
 	reg_write(isp, DCMIPP_P1HSCR, DCMIPP_P1HSCR_DEFAULT);
-	reg_write(isp, DCMIPP_P1HSSTR,
-		  DCMIPP_P1HSSTR_START(compose->width / 4, compose->height / 4));
+	reg_write(isp, DCMIPP_P1HSSTR, DCMIPP_P1HSSTR_START(0, 0));
+	/* Size (horizontal / vertical) must be multiple of the decimation */
 	reg_write(isp, DCMIPP_P1HSSZR,
-		  DCMIPP_P1HSSZR_SIZE(compose->width / 2, compose->height / 2));
+		  DCMIPP_P1HSSZR_SIZE((compose->width / 16) & ~0xf,
+				      (compose->height / 8) & ~0x07));
 }
 
 static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
